### PR TITLE
Fixes proposed for issue #389

### PR DESCRIPTION
**Pull request by smutt**
These are my proposed fixes for issue #389. I'm curious what obormot thinks of them since he/she implemented PR #267. I'm also very willing to have a discussion about this and take a different approach if necessary.

### CODE REVIEW

**Comment by smutt** [comment]
test_eth_802dot1q() is failing because pack() needs to be fixed as well. I'll work on this.

test_eth_802dot1q_stacked() is an invalid test case. Both 802.1q headers have 0x8100 as their ethertype. I'll rewrite this test.

test_eth_mpls_stacked() I'm not really sure about. I'm not familiar enough with how this is supposed to work. I'll look into it.

**Comment by obormot** [comment]
After looking at how QinQ decoding happens today it's clear we could do better. I have not tested this PR yet but I'd like to address some of the unit test changes. Specifically:

> test_eth_802dot1q_stacked() is an invalid test case. Both 802.1q headers have 0x8100 as their ethertype. I'll rewrite this test.

... not invalid;  first 0x8100 type is in the Ethernet header;  the 2nd 802.1q header has IP.  the bytes were copied from a real pcap originally found [http://packetlife.net/captures/802.1Q_tunneling.cap](here)

I've extracted a single packet used in the test 
[https://drive.google.com/open?id=12cvGfkZszTaWaFDXa_lMGrDuc-ewgoo7](here)

So we need to make sure the test passes without modifying the input data

**Comment by obormot** [comment]
I've checked a few more test cases with double 802.1Q tags, and MPLS tags, and it looks like with the proposed changes only one tag is getting decoded:

```
# this contains 2 802.1Q tags, only one gets decoded

In [11]: %paste
p1 = (
    b'\xff\xff\xff\xff\xff\xff\xca\x03\x0d\xb4\x00\x1c\x81\x00\x00\x64\x81\x00\x00\xc8\x08\x06'
    b'\x00\x01\x08\x00\x06\x04\x00\x01\xca\x03\x0d\xb4\x00\x1c\xc0\xa8\x02\xc8\x00\x00\x00\x00'
    b'\x00\x00\xc0\xa8\x02\xfe\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
)

In [12]: Ethernet(p1)
Out[12]: Ethernet(dst='\xff\xff\xff\xff\xff\xff', src='\xca\x03\r\xb4\x00\x1c', type=33024, 
vlan_tags=[VLANtag8021Q(type=2054, pri=0, id=200, cfi=0)], 
data=ARP(sha='\xca\x03\r\xb4\x00\x1c', spa='\xc0\xa8\x02\xc8', tha='\x00\x00\x00\x00\x00\x00', tpa='\xc0\xa8\x02\xfe', data='...'))

# supposed to be:
vlan_tags=[
        VLANtag8021Q(type=33024, pri=0, id=100, cfi=0), 
        VLANtag8021Q(type=2054, pri=0, id=200, cfi=0)
    ],

# the 1st tag is missing
```

File: dpkt/ethernet.py [hunk 1]
```diff
@@ -235,6 +231,42 @@ def get_type(cls, t):
     def get_type_rev(cls, k):
         return cls._typesw_rev[k]
 
+    # Property functions needed for backwards compatibility
+    def get_vlanid(self):
+        return self.vlan_tags[0].id
```
Comment (obormot) [inline dpkt/ethernet.py]: (applies to all the new properties here) indexing without checking if self.vlan_tags is empty

**Comment by smutt** [comment]
In 802.1ad the outer tag(first tag/S-tag) will typically have an ethertype of 0x88a8. That's what the IEEE standard suggests. But it needs to be something other than 0x8100 at the very least.

Wikipedia gets this right.
https://en.wikipedia.org/wiki/IEEE_802.1ad

Maybe Cisco did something different for a while in QinQ, before 802.1ad became standardized. I'm more familiar with Extreme Networks gear which used 0x9100 as default before 802.1ad was standardized, then switched over to using 0x88a8 as default.

If the ethertype of the outer tag is 0x8100 then the correct thing to do is to only decode that tag. In fact, one might even say it is a broken packet because how could a switch know that it needs to be treated differently than a normal single tagged packet? A receiving switch will need to act on the tags based on their order and ethertypes.

**Comment by obormot** [comment]
@smutt so to recap the outer tag could be 0x88a8 or 0x9100 or 0x9200 then this is double tagging and the inner tag is always 0x8100. I have found a pcap with 802.1ad frame but none of the 0x9X00 frames. 
Alternatively both tags are 0x8100 (Wikipedia mentions it as "C-tag stacking on C-tag"; legacy, but it's possible to have packets like these). I've found a couple pcaps with this, so I think dpkt should decode them both. I don't think dpkt should care about the switches' behavior (prob. different anyway, depending on the vendor), its job is to decode the packet in layers according to the structure. 
I wasn't sure if you were still interested in working on this and have spent some time to address the issue in my own PR #426 ("inspired" by this PR and the discussions.. thank you very much!). I appreciate if you take a look; alternatively we could keep iterating on this PR (fix/add unit tests etc.)

**Comment by smutt** [comment]
Hi Oscar,

Thanks for doing all this work and I think your approach makes sense for dpkt. I was maybe getting hung up on the fact that the IEEE 802.1ad standard says it MUST be 0x88a8. But I agree you're right in that dpkt SHOULD be ethertype agnostic. 

The project I was working on that lead me to suggest these changes for dpkt has somewhat fallen off my radar ATM. And to be honest I need to refamiliarize myself with the code I wrote for this PR before I can comment on your changes. I should have some time this week to look at this and get back to you. Please let me spend a bit more time on reviewing this before merging, and I promise to get back to you this month(November). Thanks again for spending your time on this.

**Comment by smutt** [comment]
Hi Oscar,

I like your changes. Please reject this PR and merge #426.

Thanks again for taking the time to work on this.

--Andrew

**Pull request closed without merging**